pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -239,73 +326,100 @@ fn storage_sync_loop<
     let remote_timelines_and_storage = Arc::new((Mutex::new(remote_timelines), remote_storage));
 
     while !crate::tenant_mgr::shutdown_requested() {
-        let next_tasks = get_next_tasks(max_concurrent_sync);
-        let remaining_queue_length = SYNC_QUEUE.len();
+        let downloads_to_sync = runtime.block_on(loop_step(
+            config,
+            Arc::clone(&remote_timelines_and_storage),
+            max_concurrent_sync,
+            max_sync_errors,
+        ));
+        // Batch timeline download registration to ensure that the external registration code won't block any running tasks before.
+        register_timeline_downloads(
+            config,
+            downloads_to_sync
+                .into_iter()
+                .map(|sync_id| (sync_id.0, sync_id.1))
+                .collect(),
+        );
+    }
 
-        if next_tasks.is_empty() {
-            trace!("No storage sync tasks found");
-            thread::sleep(Duration::from_secs(1));
-            continue;
-        } else {
-            debug!(
-                "Processing {} tasks, more tasks left to process: {}",
-                next_tasks.len(),
-                remaining_queue_length
-            );
-            REMAINING_SYNC_ITEMS.set(remaining_queue_length as i64);
+    debug!("Shutdown requested, stopping");
+    Ok(())
+}
 
-            let mut task_batch = next_tasks
-                .into_iter()
-                .map(|task| async {
-                    let sync_id = task.sync_id;
-                    let is_download = matches!(&task.kind, SyncKind::Download(_));
-                    match tokio::spawn(process_task(
-                        config,
-                        Arc::clone(&remote_timelines_and_storage),
-                        task,
-                        max_sync_errors,
-                    ))
-                    .await
-                    {
-                        Ok(()) => {
-                            if is_download {
-                                Some(sync_id)
-                            } else {
-                                None
-                            }
-                        },
-                        Err(e) => {
-                            error!(
-                                "Failed to process storage sync task for tenant {}, timeline {}: {:#}",
-                                sync_id.0, sync_id.1, e
-                            );
-                            None
-                        }
-                    }
-                })
-                .collect::<FuturesUnordered<_>>();
-
-            let mut downloads_to_sync = Vec::with_capacity(task_batch.len());
-            runtime.block_on(async {
-                while let Some(sync_id) = task_batch.next().await {
-                    if let Some(sync_id) = sync_id {
-                        downloads_to_sync.push(sync_id);
+async fn loop_step<
```

Comment:
What's the advantage of using async mpsc over sync mpsc?

Reply:
I wanted to avoid bloating the diff more, hence kept the `SYNC_QUEUE` api unchanged.
Both `Sender` and `Receiver` in `std::sync::mpsc` are `!Sync` (which is needed by the lasy init wrappers), so require their own wrappers to interact with, that would cause many changes. 
Async `sender` is fine and does not block either, so more convenient.

Also, with `std::sync::mpsc` it's a really blocking code in the async code this time, which IMO is more dangerous compared to the async counterpart (even though we call that blocking code before the main async loop call, which is not that bad with blocking calls too).

Otherwise, both sync and async versions are not that different: cancellable on drop of the pair and have bounded and unbounded flavors, so not that.
I can use the sync version, if you think it's worth it still, but not sure myself why it's beneficial.

Reply:
Interesting re `Sync`. Thanks for that insight.

I believe the earlier version had `get_next_tasks` being called from a sync context (inside `storage_sync_loop`), that's where I was thinking about using sync::mpsc.